### PR TITLE
SNOW-937665 fix unsupported column name $1

### PR DESCRIPTION
**Pull request by sfc-gh-aalam**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #SNOW-937665

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Fix a bug where `Dataframe.write.save_as_table` would throw an error for dataframe created from file api due to invalid column name `$1` error. This PR analyzes the column names to be saved and renames them in case they have name `${number}`.

### CODE REVIEW

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py [hunk 1]
```diff
@@ -570,7 +570,17 @@ def save_as_table(
         child: SnowflakePlan,
     ) -> SnowflakePlan:
         full_table_name = ".".join(table_name)
+
+        # here get the column definition from the child attributes. In certain cases we have
+        # the attributes set to ($1, VariantType()) which cannot be used as valid column name
+        # in save as table. So we rename ${number} with COL{number}.
+        hidden_column_pattern = r"\$(\d+)"
         column_definition = attribute_to_schema_string(child.attributes)
+        column_definition = re.sub(
```
Comment (sfc-gh-smirzaei) [inline src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:579]: NIT: Can we combine the two assignments?

Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:579]: I was thinking about joining it too but I decided against it to make code more readable. We build column definition in two steps. 1. converting attributes to column_definition and 2 is rename hidden column patterns into valid ones.
I am still okay with changing into a single statement.

Comment (sfc-gh-sfan) [reply src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:579]: If we want to keep them separate, should we consider giving the unprocessed variable a different name for readability, for example, `column_definition_with_hidden_columns`?

Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:579]: I like that

**Review by sfc-gh-sfan — APPROVED** [review]

**Review by sfc-gh-smirzaei — APPROVED** [review]

**Pull request merged**